collector/mdadm.go

```diff
@@ -0,0 +1,277 @@
+
+	matches := statuslineRE.FindStringSubmatch(statusline)
+
+	// +1 to make it more obvious that the whole string containing the info is also returned as matches[0].
+	if len(matches) < 3+1 {
+		return 0, 0, 0, fmt.Errorf("too few matches found in statusline: %s", statusline)
+	} else {
+		if len(matches) > 3+1 {
+			return 0, 0, 0, fmt.Errorf("too many matches found in statusline: %s", statusline)
+		}
+	}
+
+	size, err = strconv.ParseInt(matches[1], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+
+	total, err = strconv.ParseInt(matches[2], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+	active, err = strconv.ParseInt(matches[3], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+
+	return active, total, size, nil
+}
+
+// Gets the size that has already been synced out of the sync-line.
+func evalBuildline(buildline string) (int64, error) {
+	matches := buildlineRE.FindStringSubmatch(buildline)
+
+	// +1 to make it more obvious that the whole string containing the info is also returned as matches[0].
+	if len(matches) < 1+1 {
+		return 0, fmt.Errorf("too few matches found in buildline: %s", buildline)
+	}
+
+	if len(matches) > 1+1 {
+		return 0, fmt.Errorf("too many matches found in buildline: %s", buildline)
+	}
+
+	syncedSize, err := strconv.ParseInt(matches[1], 10, 64)
+
+	if err != nil {
+		return 0, fmt.Errorf("%s in buildline: %s", err, buildline)
+	}
+
+	return syncedSize, nil
+}
+
+// Parses an mdstat-file and returns a struct with the relevant infos.
+func parseMdstat(mdStatusFilePath string) ([]mdStatus, error) {
+	content, err := ioutil.ReadFile(mdStatusFilePath)
+	if err != nil {
+		return []mdStatus{}, fmt.Errorf("error parsing %s: %s", statusfile, err)
+	}
+
+	mdStatusFile := string(content)
+
+	lines := strings.Split(mdStatusFile, "\n")
+	var currentMD string
+
+	// Each md has at least the deviceline, statusline and one empty line afterwards
+	// so we will have probably something of the order len(lines)/3 devices
+	// so we use that for preallocation.
+	estimateMDs := len(lines) / 3
+	mdStates := make([]mdStatus, 0, estimateMDs)
+
+	for i, l := range lines {
+		if l == "" {
+			// Skip entirely empty lines.
+			continue
+		}
+
+		if l[0] == ' ' {
+			// Those lines are not the beginning of a md-section.
+			continue
+		}
+
+		if strings.HasPrefix(l, "Personalities") || strings.HasPrefix(l, "unused") {
+			// We aren't interested in lines with general info.
+			continue
+		}
+
+		mainLine := strings.Split(l, " ")
+		if len(mainLine) < 3 {
+			return mdStates, fmt.Errorf("error parsing mdline: %s", l)
+		}
+		currentMD = mainLine[0]               // name of md-device
+		isActive := (mainLine[2] == "active") // activity status of said md-device
+
+		active, total, size, err := evalStatusline(lines[i+1]) // parse statusline, always present
```

Comment:
The same comment as my earlier one for `mainLine` applies here - always check lengths of slices before accessing arbitrary indexes please. In this case, since you're also accessing `i+2` further down, maybe just add a single check before this like:

```go
if i+2 >= len(lines) {
  return nil, fmt.Errorf(.......)
}
```